solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -699,7 +699,10 @@ private void constructState(Set<String> changedCollections) {
           clusterState.getCollectionStates());
     }
 
-    notifyCloudCollectionsListeners();
+    if (changedCollections.isEmpty()) {
+      // We only need to notify cloud collection listeners on new collections
```

Comment:
Looks like in our case it will be `one collection` in this list. 

https://github.com/cowpaths/fullstory-solr/blob/f971df82c90e16e0ca92a5e0d9f96a3d54afd593/solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java#L1489

Reply:
Right, constructState also gets called in cases where new collections are added: https://github.com/cowpaths/fullstory-solr/blob/f971df82c90e16e0ca92a5e0d9f96a3d54afd593/solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java#L1616, but in that case it'll be an empty set.

We do want `notifyCloudCollectionsListeners()` to be called when new collections are found, which will still occur.

Reply:
I think in restart case it comes from `refreshAndWatchChildren` method, which is passing collection name?
![image](https://user-images.githubusercontent.com/8356589/214981527-6753d83f-63c8-49dc-9f80-fb4ce3dfbbe4.png)

Reply:
But, `changedCollections` will not empty right?

From this location https://github.com/cowpaths/fullstory-solr/blob/f971df82c90e16e0ca92a5e0d9f96a3d54afd593/solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java#L1489

Reply:
Correct, so we skip that call in that case when we call constructState from StateWatcher, which is what we want.

Reply:
Oh ok, Now I understand what you mean. Let me look this more closely but certainly we can improve this further. Thanks.

Reply:
The intention here is that the best possible improvement is to avoid the unnecessary long running method. We could probably improve `notifyCloudCollectionsListeners` but we shouldn't need to in order to improve restart performance, instead we can just skip it.

Reply:
I have looked all callers of `constructState` method. Following place we may miss the call. Here clusterState doesn't have collection but its calling `constructState` with collection. https://github.com/cowpaths/fullstory-solr/blob/411bf4ce97db8eb013c96ecfed1d51c349970d78/solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java#L506

Though now sure one would hit this case ever.